### PR TITLE
Chapter 14

### PR DESCRIPTION
**Pull request by mirulzaei**

### CODE REVIEW

**Comment by mirulzaei** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -1,185 +1,272 @@
-@import 'bootstrap-sprockets';
-@import 'bootstrap';
-@import 'color';
-
+@import "bootstrap-sprockets";
+@import "bootstrap";
+@import "color";
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: using '

File: app/assets/stylesheets/microposts.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Microposts controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/microposts.scss]: remove file that never used

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,4 +1,12 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   include SessionsHelper
+  private
+  def logged_in_user
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]:
```suggestion

  private
  
  def logged_in_user
```

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, :valid_user, :check_expiration, only: %i(edit update)
+
+  def new; end
+
+  def edit; end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: move this action below action create, pls

File: app/controllers/relationships_controller.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    end
+  
+    def relationship_user
+      @user = Relationship.find_by(id: params[:id]).followed
+      return if @user
+      flash[:warning] = "User not found"
+      redirect_to root_path
+    end
+  end
+
```
Comment (saobv-2223) [inline app/controllers/relationships_controller.rb:39]:
```suggestion
  end

```

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -88,3 +89,17 @@ def admin_user
...
+  @users = @user.following.page params[:page]
+  render "show_follow"
+end
+
+def followers
+  @title = "Followers"
+  @user = User.find_by id: params[:id]
+  @users = @user.followers.page params[:page]
+  render "show_follow"
+end
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: move 2 actions into block of controller, and dont forget to add endline

File: app/helpers/microposts_helper.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+module MicropostsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/microposts_helper.rb]: remove file that never used

File: app/javascript/packs/micropost.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+$(function(){
+    $('#micropost_image').bind('change', function() {
+      var size_in_megabytes = this.files[0].size/1024/1024;
+      if (size_in_megabytes > 5) { 
+        alert('Maximum file size is 5MB. Please choose a smaller file.')
+      }
+    });
+  });
```
Comment (saobv-2223) [inline app/javascript/packs/micropost.js]:
```suggestion
  });

```

**Comment by mirulzaei** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, :valid_user, :check_expiration, only: %i(edit update)
+
+  
+
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  before_action :get_user, :valid_user, :check_expiration, only: %i(edit update)

```

File: app/controllers/password_resets_controller.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+      redirect_to root_url
+    else
+      flash.now[:danger] = "Email address not found"
+      render :new
+    end
+    
+    def new; end
+
+    def edit; end
+  end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  def new; end

  def create
    @user = User.find_by email: params[:password_reset][:email].downcase

    if @user
      @user.create_reset_digest
      @user.send_password_reset_email
      flash[:info] = "Email sent with password reset instructions"
      redirect_to root_url
    else
      flash.now[:danger] = "Email address not found"
      render :new
    end
  end
  
    def edit; end
```

File: app/controllers/relationships_controller.rb [hunk 11]
```diff
@@ -0,0 +1,39 @@
+class RelationshipsController < ApplicationController
+    before_action :logged_in_user
+    before_action :load_user, only: :create
+    before_action :relationship_user, only: :destroy
+  
+  
+    def create
```
Comment (saobv-2223) [inline app/controllers/relationships_controller.rb:7]:
```suggestion
  
    def create
```

File: app/controllers/relationships_controller.rb (hunk 5 above)
Comment (saobv-2223) [inline app/controllers/relationships_controller.rb:39]:
```suggestion
  end

```
Remove last 2 spaces

File: app/javascript/packs/micropost.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+
```
Comment (saobv-2223) [inline app/javascript/packs/micropost.js]:
```suggestion
  });

```
Remove last 2 spaces

**Comment by mirulzaei** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**